### PR TITLE
Add config of DirectInputBuffer to Connector

### PR DESCRIPTION
**Pull request by FelixYBW**
#7217 introduced DirectBufferInput(DBI) to replace BufferInput (BI). Compared to BI, DBI has several advantages like fetch the data in executor as async mode, break the chunk into pieces, etc. They are controlled by 4 main parameters: loadQuantum, prefetchRowGroups, maxCoalesceBytes, maxCoalesceDistanceBytes. But the default value of these parameters aren't optimal in usage like Gluten, which fetch data rowgroup by rowgroup and didn't reuse the fetched data. As a result those default values leads to huge performance regression compared to BI. 

The PR add the confligs like loadQuantum, prefetchRowGroups, to HiveConfig so we can config them when HiveConnector is constructed. Meanwhile directorySizeGuess, filePreloadThreshold also Impacted the performance. 

- loadQuantum controls the size of each coalesce load. In Gluten, we set the default value to 256MB to make sure whole row group is fetched. Because the default rowgroup size is 128MB. If loadQuantum is smaller than rowgroup size, the rowgroup will be break as column chunks to be fetched. If it's even less than chunk size, then only the configured size is load in IO threads, all other data is loaded by task work thread in sync mode. it's why the serious performance regression happens.

- CoalesceDistance controls the size of over read. It's common that we don't select all the columns of a table. So some column chunk shouldn't be fetched which leads to non-continuous disk read. If the chunk size is smaller than coalesceDistance we can over read the data to make continuous disk read. Currently we set it 1MB in gluten

- CoalesceBytes is the unit to fetch data in IO threads, if you have many io threads in executor, you may get better performance to fetch the data in small pieces instead of whole block. CoalesceBytes controls the size. Currently we configured it as 64MB in Gluten.

- As the name indicated, prefetchRowGroups means how many rowgroup should be prefetched if engine is processing current rowgroup. Set it to more than 1 can have the overlap of next rowgroup fetching time and current rowgroup processing time. So we can get best performance.

- directorySizeGuess is renamed as footerEstimatedSize to better align with its usage.  It's used to estimate the footer data to be fetched. The footer data includs meta data of each row group.

- filePreloadThreshold  configured the threshold of file size, if a file is smaller than the threashold, we should fetch the whole file directly.

The PR is to solve the issue #8041. There are more talk in PR #7873

### CODE REVIEW

**Comment by FelixYBW** [comment]
@mbasmanova Can you take a look?

**Comment by FelixYBW** [comment]
@xiaoxmeng @majetideepak, can you help to review. The configs are essential to parquet scan performance.

@kewang1024 is the two config dynamically configured at run time? Does different session have different value?
 
`HiveConfig::isOrcUseColumnNames(const Config* session)`
`HiveConfig::isFileColumnNamesReadAsLowerCase(const Config* session)`

File: velox/connectors/hive/HiveConfig.h [hunk 1]
```diff
@@ -115,13 +115,25 @@ class HiveConfig {
   static constexpr const char* kMaxCoalescedDistanceBytes =
       "max-coalesced-distance-bytes";
 
+  /// Sets the number of prefetch rowgroups
```
Comment (mbasmanova) [inline velox/connectors/hive/HiveConfig.h:118]: drop "Sets"; add period at the end

/// The number of rowgroups to prefetch.

ditto other places.

File: velox/connectors/hive/HiveConfig.h [hunk 2]
```diff
@@ -115,13 +115,25 @@ class HiveConfig {
...
+  static constexpr const char* kPrefetchRowGroups = "prefetch-rowgroups";
+
+  /// Sets the load quantum size for a request.
```
Comment (mbasmanova) [inline velox/connectors/hive/HiveConfig.h:121]: Please, specify units.

File: velox/connectors/hive/HiveConfig.h [hunk 3]
```diff
@@ -115,13 +115,25 @@ class HiveConfig {
...
+  static constexpr const char* kLoadQuantum = "load-quantum";
+
   /// Maximum number of entries in the file handle cache.
   static constexpr const char* kNumCacheFileHandles = "num_cached_file_handles";
 
   /// Enable file handle cache.
   static constexpr const char* kEnableFileHandleCache =
       "file-handle-cache-enabled";
 
+  /// set directorySizeGuess defined in dwio::common::ReaderOptions
```
Comment (mbasmanova) [inline velox/connectors/hive/HiveConfig.h:131]: Please, provide a description that doesn't require reverse engineering the code.

File: velox/connectors/hive/HiveConfig.h [hunk 4]
```diff
@@ -115,13 +115,25 @@ class HiveConfig {
...
+  static constexpr const char* kDirectorySizeGuess = "directory-size-guess";
+
+  /// set FilePreloadThreshold defined in dwio::common::ReaderOptions
```
Comment (mbasmanova) [inline velox/connectors/hive/HiveConfig.h:134]: ditto

**Comment by FelixYBW** [comment]
> ou using previously hard-coded values as defaults?

Thank you for review. Description updated. 
Parameters are not hard-coded, #7217 use the default value. The PR adds the config to hiveConfig so gluten can config the parameters. It doesn't modify the default value.

BTW, I tested more queries from a customer side, with DBI and these parameters, compared to Vanilla spark, Gluten's performance gain increases from 1~1.3x to 2-3x depending on how serious the S3 load bottleneck is, it's really wonderful!

**Comment by mbasmanova** [comment]
@FelixYBW 

> I tested more queries from a customer side, with DBI and these parameters, compared to Vanilla spark, Gluten's performance gain increases from 1~1.3x to 2-3x depending on how serious the S3 load bottleneck is, it's really wonderful!

Happy to hear that. Thank you for sharing.

**Comment by mbasmanova** [comment]
@FelixYBW Thank you for updating PR description. It is very clear now. Note that PR description becomes a commit message on merge. Please, move the images out of PR description to the related GitHub issue. Please, address the remaining comments. Thanks.

File: velox/docs/configs.rst [hunk 5]
```diff
@@ -407,13 +407,23 @@ Each query can override the config by setting corresponding query session proper
...
-     - 512KB
+     - 128MB
      - Maximum size in bytes to coalesce requests to be fetched in a single request.
    * - max-coalesced-distance-bytes
      -
      - integer
-     - 128MB
+     - 512KB
      - Maximum distance in bytes between chunks to be fetched that may be coalesced into a single request.
+   * - load-quantum
```
Comment (majetideepak) [inline velox/docs/configs.rst:417]: `load-quantum` is repeated. Did you mean `file-preload-threshold`

File: velox/docs/configs.rst [hunk 6]
```diff
@@ -407,13 +407,18 @@ Each query can override the config by setting corresponding query session proper
    * - max-coalesced-bytes
      -
      - integer
-     - 512KB
+     - 128MB
```
Comment (mbasmanova) [inline velox/docs/configs.rst:410]: Thank you for the fix.

File: velox/docs/configs.rst [hunk 7]
```diff
@@ -436,6 +441,16 @@ Each query can override the config by setting corresponding query session proper
      - string
      - 10MB
      - Maximum bytes for sort writer in one batch of output. This is to limit the memory usage of sort writer.
+   * - file-preload-threshold
+     -
+     - integer
+     - 1MB
+     - Usually Velox fetches the meta data firstly then fetch the rest of file. But if the file is very small, Velox can fetch the whole file directly to avoid multiple IO requests. The parameter control the threshold when whole file is fetched. 
+   * - directory-size-guess
```
Comment (mbasmanova) [inline velox/docs/configs.rst:449]: > directory-size-guess

This name is a bit confusing. Which directory does this refer to? This is not a directory on the file system, is it? Wondering if we could come up with a better name here.

Comment (FelixYBW) [reply velox/docs/configs.rst:449]: According to the code, I don't see anything related to directory. Looks like it just read the following data after meta data. In dwrf, is it the directory data?

Comment (mbasmanova) [reply velox/docs/configs.rst:449]: I'm not sure. Seems like it was introduced in https://github.com/facebookincubator/velox/pull/3891 and is used in Parquet reader.

@zhenghaozzz Would you help clarify?

File: velox/docs/configs.rst [hunk 8]
```diff
@@ -436,6 +441,16 @@ Each query can override the config by setting corresponding query session proper
...
+     -
+     - integer
+     - 8MB
+     - Define the data size which is fetched with meta data together to void multiple IO requests. It's useful in file format where the immediately following data needs to be fetched after meta data
```
Comment (majetideepak) [inline velox/docs/configs.rst:453]: Can you give an example of such a file format? AFAIK both DWRF/ORC and Parquet have metadata after data.

File: velox/docs/configs.rst (hunk 7 above)
Comment (majetideepak) [reply velox/docs/configs.rst:449]: This PR only made `DIRECTORY_SIZE_GUESS` configurable. I always wondered the meaning of this as well.

**Comment by FelixYBW** [comment]
> @FelixYBW
> 
> > I tested more queries from a customer side, with DBI and these parameters, compared to Vanilla spark, Gluten's performance gain increases from 1~1.3x to 2-3x depending on how serious the S3 load bottleneck is, it's really wonderful!
> 
> Happy to hear that. Thank you for sharing.

the best configuration we used:
load-quantum = 128M
max-coalesced-bytes = 64M
max-coalesced-distance-bytes = 1M
prefetch-rowgroups = 1
directory-size-guess = 32K
file-preload-threshold = 1M
IOThreads per main task worker = 4
FLAGS_cache_prefetch_min_pct = 0  // to make sure prefetch always happen

File: velox/docs/configs.rst (hunk 7 above)
Comment (zhenghaozzz) [reply velox/docs/configs.rst:449]: @mbasmanova I agree this name is confusing but I still keep the name. It is used to guess the size of metadata in #3891.

Comment (mbasmanova) [reply velox/docs/configs.rst:449]: @zhenghaozzz Would you explain what this parameter is used for and help come up with a more descriptive name?

Comment (mbasmanova) [reply velox/docs/configs.rst:449]: Is this an estimate of how big the file footer is?

File: velox/docs/configs.rst (hunk 8 above)
Comment (FelixYBW) [reply velox/docs/configs.rst:453]: @zhenghaozzz is it for parquet only?

File: velox/docs/configs.rst (hunk 7 above)
Comment (zhenghaozzz) [reply velox/docs/configs.rst:449]: `FOOTER_SIZE_GUESS` or `FOOTER_LENGTH_GUESS`. Whatever, it guesses the size of something called footer DwrfReader or ParquetReader.

File: velox/docs/configs.rst (hunk 8 above)
Comment (zhenghaozzz) [reply velox/docs/configs.rst:453]: Both Dwrf reader and Parquet reader

File: velox/docs/configs.rst (hunk 7 above)
Comment (mbasmanova) [reply velox/docs/configs.rst:449]: Wondering if this is somehow related to StripeDictionaryCache. An internal modification to DWRF "caches" stripe metadata next to file footer to allow for fetching it in a single IO together with the footer.

CC: @HuamengJiang

File: velox/docs/configs.rst (hunk 8 above)
Comment (FelixYBW) [reply velox/docs/configs.rst:460]: Do you mean for both dwrf and parquet, we need to fetch the immediate following data after meta data? Do you know what's the data?

**Comment by FelixYBW** [comment]
@mbasmanova @majetideepak should we start a new PR to rename the directory-size-guess? Or change in this PR?

**Comment by mbasmanova** [comment]
> @mbasmanova @majetideepak should we start a new PR to rename the directory-size-guess? Or change in this PR?

Since this PR adds new config, let's make it right. We can have a follow-up PR to rename variable names to match.

**Comment by FelixYBW** [comment]
@HuamengJiang does directory-size-guess relate to StripeDictionaryCache? any name and description do you suggest?

**Comment by FelixYBW** [comment]
Since no reply from @HuamengJiang. Let's describe it as we see from code.

As @zhenghaozzz mentioned, the parameter actually is an estimation of footer size, as shown in below page

https://www.oreilly.com/library/view/operationalizing-the-data/9781492049517/ch04.html

File: velox/docs/configs.rst [hunk 9]
```diff
@@ -443,6 +448,18 @@ Each query can override the config by setting corresponding query session proper
      - string
      - 10MB
      - Maximum bytes for sort writer in one batch of output. This is to limit the memory usage of sort writer.
+   * - file-preload-threshold
+     -
+     - integer
+     - 8MB
+     - Usually Velox fetches the meta data firstly then fetch the rest of file. But if the file is very small, Velox can fetch the whole file directly to avoid multiple IO requests. 
+       The parameter controls the threshold when whole file is fetched. 
+   * - footer-size-estinated
```
Comment (mbasmanova) [inline velox/docs/configs.rst:457]: typo: estinated

Did you mean footer-size-estimate? Maybe, footer-estimated-size.

Comment (FelixYBW) [reply velox/docs/configs.rst]: Let's use  footer-estimated-size

File: velox/connectors/hive/HiveConfig.h [hunk 10]
```diff
@@ -110,20 +110,35 @@ class HiveConfig {
...
   static constexpr const char* kNumCacheFileHandles = "num_cached_file_handles";
 
   /// Enable file handle cache.
   static constexpr const char* kEnableFileHandleCache =
       "file-handle-cache-enabled";
 
+  /// The size in bytes to be fetched with Meta data together, used when the
+  /// data after meta data will be used later. Optimization to decrease small IO
+  /// request
+  static constexpr const char* kDirectorySizeGuess = "footer-estimated-sized";
```
Comment (mbasmanova) [inline velox/connectors/hive/HiveConfig.h:136]: typo: footer-estimated-sized -> footer-estimated-size

Please, update variable name and method name to match.

Comment (FelixYBW) [reply velox/connectors/hive/HiveConfig.h]: Done

**Review by mbasmanova — APPROVED** [review]
Thanks.

**Pull request closed without merging**